### PR TITLE
Fix header validation

### LINKED ISSUE #256: Panic after header download
Version: `e25fae8c736` 2022-08-10 nightly

```
2022-08-18T09:24:27.002584Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006833Z  INFO { 1/14 Headers }: Received 512 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006886Z  INFO { 1/14 Headers }: Received 4 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006895Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006905Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006913Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006926Z  INFO { 1/14 Headers }: Received 4 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006948Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006962Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006972Z  INFO { 1/14 Headers }: Received 4 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006978Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006988Z  INFO { 1/14 Headers }: Received 4 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.006994Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.007001Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.007008Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.007018Z  INFO { 1/14 Headers }: Received 4 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.007025Z  INFO { 1/14 Headers }: Received 2 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.007033Z  INFO { 1/14 Headers }: Received 1 headers from peer 0xbfa8…2a0d
2022-08-18T09:24:27.024314Z  INFO { 1/14 Headers }: Built canonical chain with=4 headers, elapsed=90.571µs
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', src/consensus/base.rs:164:47
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'thread '<unnamed><unnamed>' panicked at '' panicked at 'called `Option::unwrap()` on a `None` valuecalled `Option::unwrap()` on a `None` value', ', src/consensus/base.rssrc/consensus/base.rs::164164:47:
47
Aborted
```

### PR DESCRIPTION
**Pull request by gelfand**
Fixes #256

### CODE REVIEW

File: src/stages/headers.rs [hunk 1]
```diff
@@ -547,32 +550,34 @@ impl HeaderDownload {
...
-        if cur_size != headers.len() {
-            node.penalize_peer(peer_id).await;
-        } else {
-            let key = headers[0].1.number;
-            let last_hash = headers[headers.len() - 1].0;
 
-            let mut graph = graph.lock();
+        match Self::check_headers(&consensus, response.headers) {
+            Ok(headers) => {
+                let key = headers[0].1.number;
```
Comment (vorot93) [inline src/stages/headers.rs:564]: This will panic if `headers` empty?

File: src/consensus/base.rs [hunk 2]
```diff
@@ -152,19 +152,21 @@ impl ConsensusEngineBase {
...
-                    return Some(param::INITIAL_BASE_FEE.into());
+                    return Ok(Some(param::INITIAL_BASE_FEE.into()));
                 }
 
                 let parent_gas_target = parent.gas_limit / param::ELASTICITY_MULTIPLIER;
 
-                let parent_base_fee_per_gas = parent.base_fee_per_gas.unwrap();
+                let parent_base_fee_per_gas = parent
+                    .base_fee_per_gas
+                    .ok_or_else(|| DuoError::Validation(ValidationError::MissingBaseFee))?;
```
Comment (vorot93) [inline src/consensus/base.rs:166]: ok_or

File: src/consensus/ethash/mod.rs [hunk 3]
```diff
@@ -219,4 +207,19 @@ impl Consensus for Ethash {
             vec![]
         })
     }
+
+    fn validate_header_seal(&self, header: &BlockHeader) -> Result<(), DuoError> {
```
Comment (vorot93) [inline src/consensus/ethash/mod.rs:211]:
```suggestion
    fn validate_header_parallel(&self, header: &BlockHeader) -> Result<(), DuoError> {
```

also add `needs_parallel_validation(&self) -> bool` defaulting to `false`

File: src/stages/headers.rs [hunk 4]
```diff
@@ -631,37 +636,26 @@ impl HeaderDownload {
...
-            headers.truncate(valid_till);
+        for (i, _) in headers.iter().enumerate().skip(1) {
+            consensus.validate_block_header(&headers[i].1, &headers[i - 1].1, false)?;
         }
-        headers
+
+        Ok(headers)
     }
 
     fn verify_seal(&self, headers: &mut Vec<(H256, BlockHeader)>) {
```
Comment (vorot93) [inline src/stages/headers.rs:654]: predicate invocation of all of this on `Consensus::needs_parallel_validation` (see the relevant comment)

File: src/stages/headers.rs [hunk 5]
```diff
@@ -631,37 +636,26 @@ impl HeaderDownload {
...
-                .skip(1)
-                .enumerate()
-                .position(|(i, (hash, header))| {
-                    header.parent_hash != headers[i].0
-                        || header.number != headers[i].1.number + 1u8
-                        || node.bad_blocks.contains(hash)
-                })
-        {
-            headers.truncate(valid_till);
+        for (i, _) in headers.iter().enumerate().skip(1) {
```
Comment (vorot93) [inline src/stages/headers.rs:647]: what if headers[0] is a bad header?

**Review by vorot93 — CHANGES_REQUESTED** [review]

File: src/stages/headers.rs (hunk 1 above)
Comment (gelfand) [reply src/stages/headers.rs:564]: Yes, but we always check that `headers` are not empty, before passing them to this function.

File: src/stages/headers.rs [hunk 6]
```diff
@@ -527,7 +539,13 @@ impl HeaderDownload {
         let cur_size = headers.len();
         let took = Instant::now();
 
-        self.verify_seal(&mut headers);
+        if let Err(last_valid) = self.validate_sequentially(prev_progress_header, &headers) {
```
Comment (vorot93) [inline src/stages/headers.rs:542]: Must penalize peer here and in parallel validation

**Review by vorot93 — APPROVED** [review]

**Pull request merged**